python/trainers/buffer.py

```diff
@@ -0,0 +1,233 @@
+				# 	self[key].reset_field()
+				# 	self[key] += l
+				# self[key] = Buffer.AgentBuffer.AgentBufferField([self[key][i] for i in s])
+				# self[key].reorder(s)
+	def __init__(self):
+		self.global_buffer = self.AgentBuffer() 
+		super(Buffer, self).__init__()
+	def __str__(self):
+		return "global buffer :\n\t{0}\nlocal_buffers :\n{1}".format(str(self.global_buffer), 
+			'\n'.join(['\tagent {0} :{1}'.format(k, str(self[k])) for k in self.keys()]))
+	def __getitem__(self, key):
+		if key not in self.keys():
+			self[key] = self.AgentBuffer()
+		return super(Buffer, self).__getitem__(key)
+	def append_BrainInfo(self, info):
+		"""
+		Appends the information in a BrainInfo element to the buffer.
+		:param info: The BrainInfo from which the information will be extracted.
+		"""
+		raise BufferException("This method is not yet implemented")
+		# TODO: Find how useful this would be
+		# TODO: Implementation
+	def reset_global(self):
+		"""
+		Resets the global buffer
+		"""
+		self.global_buffer.reset_agent()
+	def reset_all(self):
+		"""
+		Resets the global buffer and all the local local_buffers
+		"""
+		# TODO: Need a better name
+		self.global_buffer.reset_agent()
+		agent_ids = list(self.keys())
+		for k in agent_ids:
+			self[k].reset_agent()
+	def append_global(self, agent_id ,key_list = None,  batch_size = None, training_length = None):
+		"""
+		Appends the buffer of an agent to the global buffer.
+		:param agent_id: The id of the agent which data will be appended
+		:param key_list: The fields that must be added. If None: all fields will be appended.
+		:param batch_size: The number of elements that must be appended. If None: All of them will be.
+		:param training_length: The length of the samples that must be appended. If None: only takes one element.
+		"""
+		if key_list is None:
+			key_list = self[agent_id].keys()
+		if not self[agent_id].check_length(key_list):
+			raise BufferException("The length of the fields {0} for agent {1} where not of comparable length"
+				.format(key_list, agent_id))
+		for field_key in key_list:
+			self.global_buffer[field_key].extend(
+				self[agent_id][field_key].get_batch(batch_size =batch_size, training_length =training_length)
+			)
+	def append_all_agent_batch_to_global(self, key_list = None,  batch_size = None, training_length = None):
+		"""
+		Appends the buffer of all agents to the global buffer.
+		:param key_list: The fields that must be added. If None: all fields will be appended.
+		:param batch_size: The number of elements that must be appended. If None: All of them will be.
+		:param training_length: The length of the samples that must be appended. If None: only takes one element.
+		"""
+		#TODO: Maybe no batch_size, only training length and a flag corresponding to "get only last of training_length"
+		for agent_id in self.keys():
+			self.append_global(agent_id ,key_list,  batch_size, training_length)
+
+
+#TODO: Put these functions into a utils class
```

Comment:
actually remove the TODO. Utils class are rarely justified because they are abused and become a soup of ambiguous functions w/o a home. Find a proper home for these functions rather than creating technical debt that utils modules typically are